File: python/quick_sort.py
def swap(arr, left, right):
    tmp = arr[left]
    arr[left] = arr[right]
    arr[right] = tmp


def partition(arr, leftBound, rightBound):
    prvot = arr[rightBound]
    left = leftBound
    right = rightBound
    while left < right:
        while left < right and arr[left] <= prvot:
            left += 1
        while left < right and arr[right] >= prvot:
            right -= 1
        if left < right:
            swap(arr, left, right)
    swap(arr, left, rightBound)
    return left


def quickSort(arr, leftBound, rightBound):
    if leftBound >= rightBound:
        return
    mid = partition(arr, leftBound, rightBound)
    quickSort(arr, leftBound, mid - 1)
    quickSort(arr, mid + 1, rightBound)


if __name__ == "__main__":
    a = [8, 3, 2, 7, 5, 9, 4, 6]
    quickSort(a, 0, len(a) - 1)
    print(a)

File: cpp/quick_sort.cpp
#include <iostream>

using namespace std;


int partition(int *ar, int l, int h) {
    int anchor = ar[l];
    while(l < h) {
        while(ar[h] >= anchor && l < h) {
            h--;
        }
        ar[l] = ar[h];
        while(ar[l] <= anchor && l < h) {
            l++;
        }
        ar[h] = ar[l];
    }
    ar[l] = anchor;
    return l;
}
void quick_sort(int *ar, int l, int h) {
    if (l < h) {
        int anchor = partition(ar, l, h);
        cout << anchor << " " << l << " " << h << endl;
        quick_sort(ar, l, anchor-1);
        quick_sort(ar, anchor+1, h);
    }
}

int main() {
    int a[5] = {5,3,1,2,4};
    quick_sort(a, 0, 4);
    for (int i = 0;i<5;++i) {
        cout << a[i] << endl;
    }
    return 0;
}